src/profiling/with_perf.md

```diff
@@ -0,0 +1,299 @@
+# Profiling with perf
+
+This is a guide for how to profile rustc with [perf](https://perf.wiki.kernel.org/index.php/Main_Page).
+
+## Initial steps
+
+- Get a clean checkout of rust-lang/master, or whatever it is you want
+  to profile.
+- Set the following settings in your `config.toml`:
+  - `debuginfo-lines = true`
+  - `use-jemalloc = false` — lets you do memory use profiling with valgrind 
+  - leave everything else the defaults
+- Run `./x.py build` to get a full build
+- Make a rustup toolchain (let's call it `rust-prof`) pointing to that result
+  - `rustup toolchain link <path-to-toolchain>`
+  
+## Gathering a perf profile
+
+perf is an excellent tool on linux that can be used to gather and
+analyze all kinds of information. Mostly it is used to figure out
+where a program spends its time. It can also be used for other sorts
+of events, though, like cache misses and so forth.
+
+### The basics
+
+The basic `perf` command is this:
+
+```
+perf record -F99 --call-graph dwarf XXX
+```
+
+The `-F99` tells perf to sample at 99 Hz, which avoids generating too
+much data for longer runs (why 99 Hz you ask? No particular reason, it
+just seems to work well for me). The `--call-graph dwarf` tells perf
```

Comment:
The usual reason given for 99 Hz is that it's unlikely to be in lockstep with any kind of periodic activity. In comparison, 100 Hz could be, and that could lead to skewed results.

Reply:
Ah, that makes sense. Thanks :)